packages/dds/tree/src/feature-libraries/forest-summary/incrementalSummaryBuilder.ts

```diff
@@ -0,0 +1,410 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils/internal";
+import type { IExperimentalIncrementalSummaryContext } from "@fluidframework/runtime-definitions/internal";
+import {
+	SummaryTreeBuilder,
+	type ReadAndParseBlob,
+} from "@fluidframework/runtime-utils/internal";
+import { setInNestedMap, tryGetFromNestedMap, type NestedMap } from "../../util/index.js";
+import type {
+	ChunkReferenceId,
+	EncodedFieldBatch,
+	IncrementalEncoderDecoder,
+	TreeChunk,
+} from "../chunked-forest/index.js";
+import type { ITreeCursorSynchronous } from "../../core/index.js";
+import { SummaryType } from "@fluidframework/driver-definitions";
+import type { IChannelStorageService } from "@fluidframework/datastore-definitions/internal";
+import type { ISnapshotTree } from "@fluidframework/driver-definitions/internal";
+import { LoggingError } from "@fluidframework/telemetry-utils/internal";
+
+/**
+ * The key for the summary blob that contains the contents of an incremental chunk.
+ */
+const chunkContentsBlobKey = "contents";
+
+/**
+ * State that tells whether a summary is currently being tracked.
+ */
+export const ForestSummaryTrackingState = {
+	/** A summary is currently being tracked. */
+	Tracking: "Tracking",
+	/** A summary is ready to be tracked. */
+	ReadyToTrack: "ReadyToTrack",
+} as const;
+export type ForestSummaryTrackingState =
+	(typeof ForestSummaryTrackingState)[keyof typeof ForestSummaryTrackingState];
+
+/**
+ * The properties of a chunk that is tracked for every summary. If a chunk doesn't change between summaries,
+ * these properties will be used to generate a summary handle for the chunk.
+ */
+interface ChunkSummaryProperties {
+	/**
+	 * The reference ID of the chunk which uniquely identifies it under its parent's summary tree.
+	 * The summary for this chunk will be stored against this reference ID as key in the summary tree.
+	 */
+	referenceId: ChunkReferenceId;
+	/**
+	 * The path for this chunk's summary in the summary tree relative to the forest's summary tree.
+	 * This path is used to generate a summary handle for the chunk if it doesn't change between summaries.
+	 */
+	summaryPath: string;
```

Comment:
can these be made read-only? I don't think we want them to be able to be edited over time.